pkg/services/ngalert/schedule/testing.go

```diff
@@ -71,6 +78,22 @@ func (f *fakeRuleStore) putRule(r *models.AlertRule) {
 	}
 }
 
+// getRecordedCommands filters recorded commands using predicate function. Returns the subset of the recorded commands that meet the predicate
+func (f *fakeRuleStore) getRecordedCommands(predicate func(cmd interface{}) (interface{}, bool)) []interface{} {
```

Comment:
(question for my own enlightment)

What is the rationale behind using handwritten fakes instead of mocks? It feels like we wouldn't have to manually maintain logic like this if we used a mock which was set up on a per-test or per-test-file basis.

Although some Go mocks require code generators, it looks like there is already precedent for generating such mocks in other packages. For instance, I see generator commands which call https://github.com/golang/mock. Is there some other historical context behind this choice?

Reply:
To be clear, I don't disagree with the use of fakes. Especially since this PR mains consistency with the existing test infrastructure. This was more to fill in context.

Reply:
I agree that we need to use mocks but the reality is that we do not do that for now. I am not aware of any decisions we made in past about mocks vs fake. Mocks just are not really popular among developers.

Reply:
Just a personal opinion here as I have used both `mockgen` and hand written fakes in the past and much prefer the latter for simple mocking. Mockgen does make more sense for more complex mocks that need expectations.

Mockgen can be slow if there are a lot of mocks to generate and it can be harder to debug errors caused when the mock is out of date (i.e. consider changing an interface and forgetting to regenerate the mocks). That said, for more complex mocks where we want expectations it makes sense to use them. I just don't think we should use them all the time for all tests. :)